src/log.rs

```diff
@@ -0,0 +1,443 @@
+
+use glib_sys;
+use once_cell::sync::Lazy;
+use std::boxed::Box as Box_;
+use std::sync::Mutex;
+use translate::*;
+use GString;
+use LogLevelFlags;
+#[cfg(any(feature = "v2_50", feature = "dox"))]
+use Variant;
+
+#[derive(Debug)]
+pub struct LogHandlerId(u32);
+
+#[doc(hidden)]
+impl FromGlib<u32> for LogHandlerId {
+    fn from_glib(value: u32) -> LogHandlerId {
+        LogHandlerId(value)
+    }
+}
+
+#[doc(hidden)]
+impl ToGlib for LogHandlerId {
+    type GlibType = u32;
+
+    fn to_glib(&self) -> u32 {
+        self.0
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum LogFlag {
+    None,
+    Recursion,
+    Fatal,
+}
+
+#[doc(hidden)]
+impl ToGlib for LogFlag {
+    type GlibType = u32;
+
+    fn to_glib(&self) -> u32 {
+        match *self {
+            LogFlag::None => 0,
+            LogFlag::Recursion => LogLevelFlags::FLAG_RECURSION.bits(),
+            LogFlag::Fatal => LogLevelFlags::FLAG_FATAL.bits(),
+        }
+    }
+}
+
+#[doc(hidden)]
+impl FromGlib<u32> for LogFlag {
+    fn from_glib(value: u32) -> LogFlag {
+        if value & LogLevelFlags::FLAG_RECURSION.bits() != 0 {
+            LogFlag::Recursion
+        } else if value & LogLevelFlags::FLAG_FATAL.bits() != 0 {
+            LogFlag::Fatal
+        } else {
+            LogFlag::None
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum LogLevel {
+    Error,
+    Critical,
+    Warning,
+    Message,
+    Info,
+    Debug,
+}
+
+#[doc(hidden)]
+impl ToGlib for LogLevel {
+    type GlibType = u32;
+
+    fn to_glib(&self) -> u32 {
+        match *self {
+            LogLevel::Error => LogLevelFlags::LEVEL_ERROR.bits(),
+            LogLevel::Critical => LogLevelFlags::LEVEL_CRITICAL.bits(),
+            LogLevel::Warning => LogLevelFlags::LEVEL_WARNING.bits(),
+            LogLevel::Message => LogLevelFlags::LEVEL_MESSAGE.bits(),
+            LogLevel::Info => LogLevelFlags::LEVEL_INFO.bits(),
+            LogLevel::Debug => LogLevelFlags::LEVEL_DEBUG.bits(),
+        }
+    }
+}
+
+#[doc(hidden)]
+impl FromGlib<u32> for LogLevel {
+    fn from_glib(value: u32) -> LogLevel {
+        if value & LogLevelFlags::LEVEL_ERROR.bits() != 0 {
+            LogLevel::Error
+        } else if value & LogLevelFlags::LEVEL_CRITICAL.bits() != 0 {
+            LogLevel::Critical
+        } else if value & LogLevelFlags::LEVEL_WARNING.bits() != 0 {
+            LogLevel::Warning
+        } else if value & LogLevelFlags::LEVEL_MESSAGE.bits() != 0 {
+            LogLevel::Message
+        } else if value & LogLevelFlags::LEVEL_INFO.bits() != 0 {
+            LogLevel::Info
+        } else if value & LogLevelFlags::LEVEL_DEBUG.bits() != 0 {
+            LogLevel::Debug
+        } else {
+            panic!("Unknown log level: {}", value)
+        }
+    }
+}
+
+#[cfg(any(feature = "v2_46", feature = "dox"))]
+pub fn log_set_handler<P: Fn(&str, LogLevel, LogFlag, &str) + Send + Sync + 'static>(
```

Comment:
The level here is a bitmask of all levels that should be allowed

Reply:
And that is a perfect demonstration of why code examples should always be provided. I completely misunderstood the docs...

Reply:
I'm only going from the documentation, so if you disagree with my interpretation let me know :)